contracts/src/LightClient.sol

```diff
@@ -0,0 +1,157 @@
+    /// @notice global storage of the finalized HotShot's light client state
+    LightClientState public finalizedState;
+    /// @notice current (finalized) epoch number
+    uint64 public currentEpoch;
+    /// @notice address of the stake table contract
+    IStakeTable internal stakeTable;
+
+    // === Data Structure ===
+    //
+    /// @notice The finalized HotShot state (as the digest of the entire HotShot state)
+    /// @param viewNum The latest view number of the finalized HotShot chain
+    /// @param blockHeight The block height of the latest finalized block
+    /// @param blockComm The block commitment (type: BN254::ScalarField)
+    /// @param feeLedgerComm The commitment to the fee ledger state (type: BN254::ScalarField)
+    /// @param stakeTableBlsKeyComm The commitment to the BlsVerKey column of the stake table
+    /// @param stakeTableSchnorrKeyComm The commitment to the SchnorrVerKey column of the table
+    /// @param stakeTableAmountComm The commitment to the stake amount column of the stake table
+    /// @param threshold The (stake-weighted) quorum threshold for a QC to be considered as valid
+    struct LightClientState {
+        uint64 viewNum;
+        uint64 blockHeight;
+        uint256 blockComm;
+        uint256 feeLedgerComm;
+        uint256 stakeTableBlsKeyComm;
+        uint256 stakeTableSchnorrKeyComm;
+        uint256 stakeTableAmountComm;
+        uint256 threshold;
+    }
+
+    /// @notice Event that a new finalized state has been successfully verified and updated
+    event NewState(uint64 indexed viewNum, uint64 indexed blockHeight, uint256 blockComm);
+
+    /// @notice The state is outdated and older than currently known `finalizedState`
+    error OutdatedState();
+    /// @notice Warning that the last block of the current epoch is not yet submitted before newer
+    /// states are proposed.
+    error MissingLastBlockForCurrentEpoch(uint64 expectedBlockHeight);
+    /// @notice Invalid user inputs: wrong format or non-sensible arguments
+    error InvalidArgs();
+
+    constructor(LightClientState memory genesis, address stakeTableAddr) {
+        if (genesis.viewNum != 0 || genesis.blockHeight != 0) {
+            revert InvalidArgs();
+        }
+
+        genesisState = genesis;
+        finalizedState = genesis;
+        currentEpoch = 0;
+        stakeTable = IStakeTable(stakeTableAddr);
+        // TODO: (alex) initialized stake table or at least store its contract address ref here
+    }
+
+    // === State Modifying APIs ===
+    //
+    /// @notice Update the latest finalized light client state. It is expected to be updated
+    /// periodically, especially an update for the last block for every epoch has to be submitted
+    /// before any newer state can be accepted since the stake table commitments of that block
+    /// become the snapshots used for vote verifications later on.
+    function newFinalizedState(
+        LightClientState calldata newState,
```

Comment:
```suggestion
        LightClientState memory newState,
```

IIRC it's cheaper in terms of gas if the data is loaded into memory in the function.